File: src/mk2/JavaPackage/src/main/java/tamaf/FIPARequestResponderForDispatcher.java
package tamaf;

public class FIPARequestResponderForDispatcher extends FSMBehavior {
    protected ACLMessage requestMessage;

    public static final int TRANSITION_AGREED = 0;
    public static final int TRANSITION_REFUSED = 1;
    public static final int TRANSITION_DONE = 0;

    public FIPARequestResponderForDispatcher(Agent agent, ACLMessage requestMessage) {
        super(agent);
        this.requestMessage = requestMessage;

        addInitialState(new HandleRequestBehavior(this), "handleRequest");
        addState(new PrepareResultBehavior(this), "prepareResult");
        addFinalState(new OneShotBehavior(agent) { @Override public void action() {} }, "end");

        addTransition("handleRequest", "prepareResult", TRANSITION_AGREED);
        addTransition("handleRequest", "end", TRANSITION_REFUSED);
        addTransition("prepareResult", "end", TRANSITION_DONE);
    }

    public ACLMessage handleRequest(ACLMessage request) { return null; }
    public ACLMessage prepareResultNotification(ACLMessage request, ACLMessage response) { return null; }

    private class HandleRequestBehavior extends SimpleBehavior {
        FIPARequestResponderForDispatcher fsm;
        int transitionCode = -1;
        boolean isDone = false;

        HandleRequestBehavior(FIPARequestResponderForDispatcher fsm) { super(fsm.getAgent()); this.fsm = fsm; }
        @Override
        public void action() {
            ACLMessage reply = fsm.handleRequest(fsm.requestMessage);
            if (reply == null) return;

            if (fsm.requestMessage.getConversationId() != null) reply.setConversationId(fsm.requestMessage.getConversationId());
            fsm.sharedData.put("agreeMessage", reply);
            fsm.getAgent().send(reply);
            transitionCode = (reply.getPerformative() == Performative.AGREE) ? TRANSITION_AGREED : TRANSITION_REFUSED;
            isDone = true;
        }
        @Override
        public boolean done() { return isDone; }
        @Override
        public int onEnd() { return transitionCode; }
    }

    private class PrepareResultBehavior extends SimpleBehavior {
        FIPARequestResponderForDispatcher fsm;
        boolean isDone = false;

        PrepareResultBehavior(FIPARequestResponderForDispatcher fsm) { super(fsm.getAgent()); this.fsm = fsm; }
        @Override
        public void action() {
            ACLMessage resultMsg = fsm.prepareResultNotification(fsm.requestMessage, (ACLMessage) fsm.sharedData.get("agreeMessage"));
            if (resultMsg == null) return;

            if (fsm.requestMessage.getConversationId() != null) resultMsg.setConversationId(fsm.requestMessage.getConversationId());
            fsm.getAgent().send(resultMsg);
            isDone = true;
        }
        @Override
        public boolean done() { return isDone; }
    }
}

File: src/mk1/PythonPackage/src/tamaf/fiparequestresponderfordispatcher.py
from typing import TYPE_CHECKING
from .fsmbehavior import FSMBehavior
from .simplebehavior import SimpleBehavior
from .performative import Performative

if TYPE_CHECKING:
    from .agent import Agent
    from .aclmessage import ACLMessage

class FIPARequestResponderForDispatcher(FSMBehavior):
    
    TRANSITION_AGREED = 0
    TRANSITION_REFUSED = 1
    TRANSITION_DONE = 0
    
    def __init__(self, agent: 'Agent', requestMessage: 'ACLMessage'):
        super().__init__(agent)
        self.requestMessage = requestMessage

        class HandleRequestBehavior(SimpleBehavior):
            def __init__(self, fsm: 'FIPARequestResponderForDispatcher'):
                super().__init__(fsm.agent)
                self.fsm = fsm
                self.transitionCode = -1
                self.is_done = False

            def Action(self):
                reply = self.fsm.handleRequest(self.fsm.requestMessage)
                if reply is None:
                    return

                if getattr(self.fsm.requestMessage, 'conversationid', None):
                    reply.conversationid = self.fsm.requestMessage.conversationid
                
                self.fsm.sharedData['agreeMessage'] = reply
                self.agent.Send(reply)
                
                if reply.performative == Performative.AGREE:
                    self.transitionCode = self.fsm.TRANSITION_AGREED
                else:
                    self.transitionCode = self.fsm.TRANSITION_REFUSED
                
                self.is_done = True

            def Done(self):
                return self.is_done

            def OnEnd(self):
                return self.transitionCode

        class PrepareResultBehavior(SimpleBehavior):
            def __init__(self, fsm: 'FIPARequestResponderForDispatcher'):
                super().__init__(fsm.agent)
                self.fsm = fsm
                self.is_done = False

            def Action(self):
                resultMsg = self.fsm.prepareResultNotification(
                    self.fsm.requestMessage, 
                    self.fsm.sharedData['agreeMessage']
                )
                if resultMsg is None:
                    return

                if getattr(self.fsm.requestMessage, 'conversationid', None):
                    resultMsg.conversationid = self.fsm.requestMessage.conversationid
                self.agent.Send(resultMsg)
                self.is_done = True

            def Done(self):
                return self.is_done

            def OnEnd(self):
                return self.fsm.TRANSITION_DONE

        class EndResponderBehavior(OneShotBehavior):
            def Action(self): pass 

        self.AddInitialState(HandleRequestBehavior(self), "handleRequest")
        self.AddState(PrepareResultBehavior(self), "prepareResult")
        self.AddFinalState(EndResponderBehavior(self), "end")

        self.AddTransition("handleRequest", "prepareResult", self.TRANSITION_AGREED)
        self.AddTransition("handleRequest", "end", self.TRANSITION_REFUSED)
        self.AddTransition("prepareResult", "end", self.TRANSITION_DONE)

    def handleRequest(self, requestMessage: 'ACLMessage') -> 'ACLMessage':
        pass
    def prepareResultNotification(self, requestMessage: 'ACLMessage', responseMessage: 'ACLMessage') -> 'ACLMessage':
        pass
